tests/integration/ha_tests/test_scaling_three_units.py

```diff
@@ -0,0 +1,135 @@
+    # This is a potentially destructive test, so it shouldn't be run against existing clusters
+    charm = await ops_test.build_charm(".")
+    async with ops_test.fast_forward():
+        # Deploy the first cluster with reusable storage
+        await gather(
+            ops_test.model.deploy(
+                charm,
+                application_name=DATABASE_APP_NAME,
+                num_units=3,
+                base=CHARM_BASE,
+                config={"profile": "testing"},
+            ),
+            ops_test.model.deploy(
+                APPLICATION_NAME,
+                application_name=APPLICATION_NAME,
+                base=CHARM_BASE,
+                channel="edge",
+            ),
+        )
+
+        await ops_test.model.wait_for_idle(status="active", timeout=1500)
+
+
+@pytest.mark.group(1)
+@markers.juju3
+@pytest.mark.parametrize(
+    "roles",
+    [
+        ["primaries"],
+        ["sync_standbys"],
+        ["replicas"],
+        ["primaries", "replicas"],
+        ["sync_standbys", "replicas"],
+    ],
+)
+@pytest.mark.abort_on_fail
+async def test_removing_unit(ops_test: OpsTest, roles: list[str], continuous_writes) -> None:
+    logger.info(f"removing {', '.join(roles)}")
+    # Start an application that continuously writes data to the database.
+    app = await app_name(ops_test)
+    original_roles = await get_cluster_roles(
+        ops_test, ops_test.model.applications[DATABASE_APP_NAME].units[0].name
+    )
+    await start_continuous_writes(ops_test, app)
+    units = [original_roles[role][0] for role in roles]
+    for unit in units:
+        logger.info(f"Stopping unit {unit}")
+        await stop_machine(ops_test, await get_machine_from_unit(ops_test, unit))
+    await sleep(15)
+    for unit in units:
+        logger.info(f"Deleting unit {unit}")
+        await ops_test.model.destroy_unit(unit, force=True, destroy_storage=False, max_wait=1500)
+
+    if len(roles) > 1:
+        for left_unit in ops_test.model.applications[DATABASE_APP_NAME].units:
+            if left_unit.name not in units:
+                break
+
+        try:
+            await ops_test.model.block_until(
+                lambda: left_unit.workload_status == "blocked"
+                and left_unit.workload_status_message
+                == "Raft majority loss, run: promote-to-primary",
+                timeout=600,
+            )
+
+            run_action = (
+                await ops_test.model.applications[DATABASE_APP_NAME]
+                .units[0]
+                .run_action("promote-to-primary", scope="unit", force=True)
+            )
+            await run_action.wait()
+        except exceptions.TimeoutError:
+            # Check if Patroni self healed
+            assert (
+                left_unit.workload_status == "active"
+                and left_unit.workload_status_message == "Primary"
+            )
+            logger.warning(f"Patroni self-healed without raft reinitialisation for roles {roles}")
```

Comment:
Sometimes when removing the primary and async replica, Patroni manages to survive, so adding an exception for this case. Should I nail it down further?